File: app/registry/dockerhub.py
from typing import Optional

import requests

from app.config import log
import app.http as _http


def get_dockerhub_token(username: str, password: str) -> Optional[str]:
    """
    Exchange DockerHub credentials for a JWT token.
    Both plain passwords and Personal Access Tokens (PATs) are accepted as `password`.
    PATs are recommended — generate one at:
    https://app.docker.com/settings/personal-access-tokens
    """
    if not username or not password:
        log.info("DockerHub: no credentials — using anonymous access (rate-limited).")
        return None
    try:
        resp = _http.http_session.post(
            "https://hub.docker.com/v2/users/login",
            json={"username": username, "password": password},
            timeout=15,
        )
        resp.raise_for_status()
        token = resp.json().get("token")
        log.info("DockerHub authentication successful.")
        return token
    except Exception as exc:
        log.warning(f"DockerHub auth failed: {exc}. Falling back to anonymous access.")
        return None


def _fetch_dockerhub_tags(image_name: str, token: Optional[str], current_tag: Optional[str] = None) -> list[str]:
    """Fetch tags from Docker Hub using the Hub v2 API.

    Docker Hub returns tags ordered by last_updated descending (newest first).
    When *current_tag* is given we stop paginating once the current tag appears
    on a page, because all subsequent pages contain only older entries.
    """
    # Strip explicit docker.io/ prefix if present
    name = image_name.removeprefix("docker.io/")
    parts = name.split("/")
    namespace = "library" if len(parts) == 1 else parts[0]
    repo = parts[-1]

    headers = {"Authorization": f"Bearer {token}"} if token else {}
    tags: list[str] = []
    url = f"https://hub.docker.com/v2/namespaces/{namespace}/repositories/{repo}/tags?page_size=100&ordering=last_updated"

    while url:
        try:
            resp = _http.http_session.get(url, headers=headers, timeout=20)
            resp.raise_for_status()
            data = resp.json()
            page_tags = [t["name"] for t in data.get("results", [])]
            tags.extend(page_tags)

            # Early stop: current tag found on this page → older pages not needed
            if current_tag and current_tag in page_tags:
                log.debug(f"DockerHub: found current tag '{current_tag}' — stopping pagination")
                break

            url = data.get("next")
        except requests.HTTPError as exc:
            log.error(f"DockerHub HTTP error for {namespace}/{repo}: {exc}")
            break
        except Exception as exc:
            log.error(f"DockerHub error for {namespace}/{repo}: {exc}")
            break

    return tags

File: src/registry/dockerhub.rs
use reqwest::Client;
use serde::Deserialize;

#[derive(Deserialize)]
struct LoginResponse {
    token: Option<String>,
}

#[derive(Deserialize)]
struct TagsPage {
    results: Vec<TagEntry>,
    next: Option<String>,
}

#[derive(Deserialize)]
struct TagEntry {
    name: String,
}

pub async fn get_dockerhub_token(
    client: &Client,
    username: &str,
    password: &str,
    login_url: &str,
) -> Option<String> {
    if username.is_empty() || password.is_empty() {
        return None;
    }
    let resp = client
        .post(login_url)
        .json(&serde_json::json!({"username": username, "password": password}))
        .send()
        .await
        .ok()?;
    if !resp.status().is_success() {
        return None;
    }
    let data: LoginResponse = resp.json().await.ok()?;
    data.token
}

pub async fn fetch_dockerhub_tags(
    client: &Client,
    image: &str,
    token: Option<&str>,
    current_tag: Option<&str>,
    api_base: &str,
) -> Vec<String> {
    let name = image.strip_prefix("docker.io/").unwrap_or(image);
    let parts: Vec<&str> = name.split('/').collect();
    let namespace = if parts.len() == 1 {
        "library"
    } else {
        parts[0]
    };
    let repo = parts[parts.len() - 1];

    let mut url = format!(
        "{api_base}/v2/namespaces/{namespace}/repositories/{repo}/tags?page_size=100&ordering=last_updated"
    );
    let mut tags = Vec::new();

    loop {
        let mut req = client.get(&url);
        if let Some(t) = token {
            req = req.header("Authorization", format!("Bearer {t}"));
        }
        let resp = match req.send().await {
            Ok(r) if r.status().is_success() => r,
            _ => break,
        };
        let page: TagsPage = match resp.json().await {
            Ok(p) => p,
            Err(_) => break,
        };
        let next_url = page.next.clone();
        let page_tags: Vec<String> = page.results.into_iter().map(|t| t.name).collect();
        let found = current_tag
            .map(|ct| page_tags.iter().any(|t| t == ct))
            .unwrap_or(false);
        tags.extend(page_tags);
        if found {
            break;
        }
        match next_url {
            Some(u) if !u.is_empty() => url = u,
            _ => break,
        }
    }

    tags
}
